File: Python/arraylist/arraylist.py
class ArrayList:
	INITIAL_CAPACITY = 16
	
	def __init__(self, elems = None):
		self._capacity = INITIAL_CAPACITY
		self._data = [None] * self._capacity
		self._length = 0
		if elems:
			for e in elems:
				self.add(e)
	
	def __len__(self):
		return self._length
	
	def __getitem__(self, index):
		return self._data[index]
	
	def __setitem__(self, index, e):
		elf._data[index] = e
	
	def __contains__(self, e):
		return self.contains(e)
	
	@property
	def capacity(self):
		return self._capacity
	
	def empty(self):
		return self._length == 0
	
	def get(self, index):
		return self._data[index]
	
	def set(self, index, e):
		temp = self._data[index]
		self._data[index] = e
		return temp
	
	def contains(self, e):
		for i in self._data:
			if i == key:
				return True
		return False
	
	def index(self, e):
		for i, v in enumerate(self.data):
			if v == e:
				return i
		return -1
	
	def add(self, e):
		if self._capacity == self._length:
			self._capacity *= 2
			temp = [None] * self._capacity
			for i in range(self._length):
				temp[i] = self._data[i]
			self._data = temp
		self._data[self._length] = e
		self._length += 1
	
	def insert(self, index, e):
		if self._capacity == self._length:
			self._capacity *= 2
			temp = [None] * self._capacity
			for i in range(self._length):
				temp[i] = self._data[i]
			self._data = temp
		for i in range(self._length, index - 1, -1):
			self._data[i] = self._data[i - 1]
		self._length += 1
		self._data[index] = e
	
	def remove(self, index):
		temp = self.data[index]
		self._length -= 1
		for i in range(index, self._length):
			self._data[i] = self._data[i + 1]
		return temp
	
	def erase(self, e):
		i = self.index(e)
		if i >= 0:
			self.remove(i)
			return True
		return False
	
	def clear(self):
		self._capacity = INITIAL_CAPACITY
		self._data = [None] * self._capacity
		self._length = 0

File: Ruby/arraylist/arraylist.rb
class ArrayList	
	INITIAL_CAPACITY = 16
	attr_reader :size, :capacity
	
	def initalize(elems = nil)
		@size = 0
		@capacity = INITIAL_CAPACITY
		@data = Array.new(INITIAL_CAPACITY)
		unless elems.nil?
			elems.each { |e| add(e) }
		end
	end
	
	def [](index)
		@data[index]
	end
	
	def []=(index, e)
		@data[index] = e
	end
	
	def length
		@size
	end
	
	def empty?
		@size == 0
	end
	
	def get(index)
		@data[index]
	end
	
	def set=(index, e)
		@data[index] = e
	end
	
	def contains(e)
		@data.each { |v| return true if v == e }
		false
	
	def index(e)
		@data.each_with_index { |v, i| return i if v == e }
		-1
	end
	
	def add(e)
		if @size == @capacity
			@capacity *= 2
			temp = Array.new(@capacity)
			for i in 0...@size
				temp[i] = @data[i]
			end
			@data = temp
		end
		@data[@size] = e
		@size += 1
	end
	
	def insert(index, e)
		if @size == @capacity
			@capacity *= 2
			temp = Array.new(@capacity)
			for i in 0...@size
				temp[i] = @data[i]
			end
			@data = temp
		end
		@size.downto(@index) { |i| @data[i] = @data[] }
		@data[index] = e
		@size += 1
	end
	
	def remove(index)
		temp = @data[index]
		@size -= 1
		(index..@size).each { |i| @data[i] = @data[i + 1] }
		temp
	end
	
	def erase(e)
		i = index(e)
		if i >= 0
			remove(i)
			return true
		end
		false
	end
	
	def clear
		@capacity = INITIAL_CAPACITY
		@data = Array.new(INITIAL_CAPACITY)
		@size = 0
	end
end
